### PR TITLE
fix allowing assigning JSON fields back to NULL

### PR DESCRIPTION
**Pull request by GeoffreyPlitt**
Fixes https://github.com/sequelize/sequelize/issues/3877

### CODE REVIEW

**Comment by alexcarruthers** [comment]
What's the timeline for putting this into a release? I'm in the midst of updating from an older version and this is one of the few things that's blocking me and I'd rather not be on one specific commit again.

**Comment by Chris911** [comment]
We really need this to move forward with our integration of Sequelize. Please merge/release as soon as possible.

**Comment by janmeier** [comment]
Needs a unit test before we can merge

**Comment by Chris911** [comment]
@GeoffreyPlitt let me know if you need help with unit testing.

**Comment by GeoffreyPlitt** [comment]
I'm getting a weird error, any ideas? Seems like lodash is null or something. I did "npm install".

TypeError: undefined is not a function
    at /home/ubuntu/ExpoSync/node_modules/sequelize/lib/utils.js:99:21

File: test/integration/model/json.test.js [hunk 1]
```diff
@@ -243,6 +243,22 @@ describe(Support.getTestDialectTeaser('Model'), function() {
...
       });
 
+      it('should be possible to save/update a NULL as the field value', function () {
+        return this.Event.create({data:{a:1}}) // create with non-NULL data field
+        .then( function(event) { // set data field to null
+          event.data = null;
+          return event.save();
+        })
+        .then( function(event) { // fetch it again
+          return this.Event.findAll({id: event.id});
```
Comment (BridgeAR) [inline test/integration/model/json.test.js]: "this" refers to the inner function and therefor does not know about the Event. You have to either bind this to the promise or add a variable to hold this as in:
```js
var self = this;
```

**Comment by GeoffreyPlitt** [comment]
Cool yeah that fix should be fine. I'm still getting this problem though when I use this in my other project, any ideas?

TypeError: undefined is not a function
at /home/ubuntu/ExpoSync/node_modules/sequelize/lib/utils.js:99:21

**Comment by GeoffreyPlitt** [comment]
@BridgeAR Can you take a look? I'm using your self=this trick (I called it "that") but still no good.

File: test/integration/model/json.test.js [hunk 2]
```diff
@@ -243,6 +243,23 @@ describe(Support.getTestDialectTeaser('Model'), function() {
         });
       });
 
+      it('should be possible to save/update a NULL as the field value', function () {
+        var that = this
```
Comment (BridgeAR) [inline test/integration/model/json.test.js]: Missing semikolon^^

File: test/integration/model/json.test.js [hunk 3]
```diff
@@ -243,6 +243,23 @@ describe(Support.getTestDialectTeaser('Model'), function() {
...
+          event.data = null;
+          return event.save();
+        })
+        .then( function(event) { // fetch it again
+          return that.Event.findAll({id: event.id});
+        })
+        .then( function(events) { // asserts
+          expect(events.length).to.equal(1);
+          var new_event = events[0];
+          expect(new_event.get('data').to.eql(null));
```
Comment (BridgeAR) [inline test/integration/model/json.test.js]: The bracket is not set right. It should be `expect(new_event.get('data')).to.eql(null);`

**Comment by GeoffreyPlitt** [comment]
Can someone with write access cancel the Travis job for my old (not latest) commit? It seems to be blocking the new one from running.

**Comment by GeoffreyPlitt** [comment]
Cancel this one: https://travis-ci.org/sequelize/sequelize/builds/66610821

**Comment by GeoffreyPlitt** [comment]
@BridgeAR @Chris911 Can one of you cancel the older commit Travis build?

**Comment by GeoffreyPlitt** [comment]
@BridgeAR @Chris911 It appears that my test passes, and there's an error about deadlocks that seems unrelated.

**Comment by GeoffreyPlitt** [comment]
Squashed.

**Comment by janmeier** [comment]
Could you add a unit test that just test that the value is correctly escaped instead [here](https://github.com/sequelize/sequelize/blob/master/test/unit/sql/json.test.js)

There is no need to add an integration test which has to hit the database to test this.

File: lib/utils.js [hunk 4]
```diff
@@ -96,7 +96,7 @@ var Utils = module.exports = {
   cloneDeep: function(obj, fn) {
     return lodash.cloneDeep(obj, function (elem) {
       // Preserve special data-types like `fn` across clones. _.get() is used for checking up the prototype chain
-      if (!!Utils._.get(elem, 'clone') && typeof elem.clone === 'function') {return elem.clone(); }
+      if (Utils._ && Utils._.get && !!Utils._.get(elem, 'clone') && typeof elem.clone === 'function') {return elem.clone(); }
```
Comment (janmeier) [inline lib/utils.js]: Hmm, seems there must be something wrong with the version of lodash you have installed if this fails.

But just checking `elem && typeof elem.clone` should actually be enough here

**Comment by alexcarruthers** [comment]
@GeoffreyPlitt are you hoping to get this landed today?

**Comment by GeoffreyPlitt** [comment]
Yeah I'll look at the notes above and revise my tests within the next few
hours.

On Mon, Jun 15, 2015 at 9:17 AM, Alex Carruthers <notifications@github.com>
wrote:

> @GeoffreyPlitt <https://github.com/GeoffreyPlitt> are you hoping to get
> this landed today?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sequelize/sequelize/pull/3923#issuecomment-112124702>.
>



-- 
http://www.geoffplitt.com
http://facebook.com/geoffrey.plitt
https://twitter.com/GeoffreyPlitt
773.339.0915

**Comment by GeoffreyPlitt** [comment]
Cleaned up and squashed. @janmeier I think I did what you suggested, can you take a look?

**Comment by GeoffreyPlitt** [comment]
To the best of my knowledge this is ready to merge.

**Comment by GeoffreyPlitt** [comment]
I had to fix one more thing. It seems that cloneDeep in lib/utils.js is being called with a model for some reason, and you can't clone a model (circular structures). I don't know why the model is being cloned but this seems to fix it.

**Comment by GeoffreyPlitt** [comment]
squashed again.

File: lib/utils.js [hunk 5]
```diff
@@ -96,7 +96,11 @@ var Utils = module.exports = {
   cloneDeep: function(obj, fn) {
     return lodash.cloneDeep(obj, function (elem) {
       // Preserve special data-types like `fn` across clones. _.get() is used for checking up the prototype chain
-      if (!!Utils._.get(elem, 'clone') && typeof elem.clone === 'function') {return elem.clone(); }
+      try {
+        if (!!Utils._.get(elem, 'clone') && typeof elem.clone === 'function') {return elem.clone(); }
+      } catch(ex) {
+        return elem;
+      }
```
Comment (mickhansen) [inline lib/utils.js]: A try/catch should really not be needed. It also throws the code into deopt.

**Comment by janmeier** [comment]
Merged in https://github.com/sequelize/sequelize/commit/0cee002ab173aa83f0db7ee16571b9bcc9cc8f32 (in removed the try manaully)

**Comment by GeoffreyPlitt** [comment]
Early warning that I think something is wrong after this merge. My symptoms are that complex models involving `belongsTo()` relationships, when queried with `include: ...` arguments, break at `lib/utils.js:99` due to `cloneDeep()` errors. Looking for more definitive proof now.

**Comment by GeoffreyPlitt** [comment]
@FredKSchott since you authored https://github.com/sequelize/sequelize/commit/08ed3756e5102214d3b844f2d655ab80c4adede0 which is where this first started happening for me (I just did git-bisect), do you have any insight into why circular structures are being passed into cloneDeep()?  Here's a full stack-trace.

```bash
Error: Can't clone [object SequelizeModel:Expos]
    at Function.Object.clone (/vagrant/node_modules/q-io/node_modules/collections/shim-object.js:509:27)
    at Function.Object.clone (/vagrant/node_modules/q-io/node_modules/collections/shim-object.js:506:45)
    at Array.<anonymous> (/vagrant/node_modules/q-io/node_modules/collections/shim-array.js:249:31)
    at /vagrant/node_modules/sequelize/lib/utils.js:99:90
    at baseClone (/vagrant/node_modules/sequelize/node_modules/lodash/index.js:1693:27)
    at /vagrant/node_modules/sequelize/node_modules/lodash/index.js:1738:23
    at /vagrant/node_modules/sequelize/node_modules/lodash/index.js:3106:15
    at baseForOwn (/vagrant/node_modules/sequelize/node_modules/lodash/index.js:2053:14)
    at baseClone (/vagrant/node_modules/sequelize/node_modules/lodash/index.js:1737:39)
    at Function.cloneDeep (/vagrant/node_modules/sequelize/node_modules/lodash/index.js:8345:11)
    at Object.module.exports.cloneDeep (/vagrant/node_modules/sequelize/lib/utils.js:97:19)
    at Model.__optClone (/vagrant/node_modules/sequelize/lib/model.js:306:16)
    at Model.findAll (/vagrant/node_modules/sequelize/lib/model.js:1157:13)
```

**Comment by janmeier** [comment]
@GeoffreyPlitt The stack-trace seems to indicate an error in https://github.com/montagejs/collections/blob/master/shim-object.js - the collections module adds a `clone` function to object.prototype.

Sequelize expects a clone function on its own internal objects that can clone them. But appearently the clone function added by collections cannot clone a sequelize model. 

We are deep cloning when cloning an options object - the reason we are try to clone a sequelize model instance is because we are cloning an include.

Generally, sequelize does not play well with modules that alter object / array prototypes

**Comment by GeoffreyPlitt** [comment]
Cool, great insight into the problem, thanks. I'd like to write the missing
test case to produce this, i.e. a scenario that combines the cloning step
with a find-with-include query... I just got the dockers built, so I may be
able to do it. Any tips where this would go? I messed up last time when I
picked a place/level to add a test.

On Wed, Jun 17, 2015 at 12:17 AM, Jan Aagaard Meier <
notifications@github.com> wrote:

> @GeoffreyPlitt <https://github.com/GeoffreyPlitt> The stack-trace seems
> to indicate an error in
> https://github.com/montagejs/collections/blob/master/shim-object.js - the
> collections module adds a clone function to object.prototype.
>
> Sequelize expects a clone function on its own internal objects that can
> clone them. But appearently the clone function added by collections cannot
> clone a sequelize model.
>
> We are deep cloning when cloning an options object - the reason we are try
> to clone a sequelize model instance is because we are cloning an include.
>
> Generally, sequelize does not play well with modules that alter object /
> array prototypes
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sequelize/sequelize/pull/3923#issuecomment-112686913>.
>



-- 
http://www.geoffplitt.com
http://facebook.com/geoffrey.plitt
https://twitter.com/GeoffreyPlitt
773.339.0915

**Comment by janmeier** [comment]
Hmm, I guess it would fit in include.test or models/include.test. Or perhaps you could write a unit test that only tests the clone function.

But note that your unit-test will only fail if object.prototype has a `clone` method...

We do actually try to avoid cloning sequelize models, [here](https://github.com/sequelize/sequelize/blob/master/lib/model.js#L304). The problem is that the code in utils.cloneDeep is run _before_ we check if the object is a sequelize model - so we try to invoke the clone  function.

Perhaps it would make sense to merge those two functions (utils.cloneDeep and model.optClone) - the problem is that we don't have access to sequelize.model in utils - that would create a circular dependency. 

If you can find a way to merge the two functions so we check for elem instanceof sequelize.model before trying the clone function, we would probably accept that PR :)

**Pull request closed without merging**